common/cuda_hip/distributed/matrix_kernels.hpp.inc

```diff
@@ -0,0 +1,271 @@
+                        input.get_const_row_idxs(),
+                        input.get_const_row_idxs() + num_input_elements,
+                        row_range_ids.get_data());
+    array<size_type> col_range_ids{exec, input.get_num_elems()};
+    thrust::upper_bound(thrust::device, col_range_bounds + 1,
+                        col_range_bounds + num_col_ranges + 1,
+                        input.get_const_col_idxs(),
+                        input.get_const_col_idxs() + num_input_elements,
+                        col_range_ids.get_data());
+
+    // count number of diag<0> and offdiag<1> elements
+    auto range_ids_it = thrust::make_zip_iterator(thrust::make_tuple(
+        row_range_ids.get_const_data(), col_range_ids.get_const_data()));
+    auto num_elements_pair = thrust::transform_reduce(
+        thrust::device, range_ids_it, range_ids_it + num_input_elements,
+        [local_part, row_part_ids, col_part_ids] __host__ __device__(
+            const thrust::tuple<size_type, size_type>& tuple) {
+            auto row_part = row_part_ids[thrust::get<0>(tuple)];
+            auto col_part = col_part_ids[thrust::get<1>(tuple)];
+            bool is_inner_entry =
+                row_part == local_part && col_part == local_part;
+            bool is_ghost_entry =
+                row_part == local_part && col_part != local_part;
+            return thrust::make_tuple(
+                is_inner_entry ? size_type{1} : size_type{0},
+                is_ghost_entry ? size_type{1} : size_type{0});
+        },
+        thrust::make_tuple(size_type{}, size_type{}),
+        [] __host__ __device__(const thrust::tuple<size_type, size_type>& a,
+                               const thrust::tuple<size_type, size_type>& b) {
+            return thrust::make_tuple(thrust::get<0>(a) + thrust::get<0>(b),
+                                      thrust::get<1>(a) + thrust::get<1>(b));
+        });
+    auto num_diag_elements = thrust::get<0>(num_elements_pair);
+    auto num_offdiag_elements = thrust::get<1>(num_elements_pair);
+
+    // define global-to-local maps for row and column indices
+    auto map_to_local_row =
+        [row_range_bounds, row_range_starting_indices] __host__ __device__(
+            const GlobalIndexType row, const size_type range_id) {
+            return static_cast<LocalIndexType>(row -
+                                               row_range_bounds[range_id]) +
+                   row_range_starting_indices[range_id];
+        };
+    auto map_to_local_col =
+        [col_range_bounds, col_range_starting_indices] __host__ __device__(
+            const GlobalIndexType col, const size_type range_id) {
+            return static_cast<LocalIndexType>(col -
+                                               col_range_bounds[range_id]) +
+                   col_range_starting_indices[range_id];
+        };
+
+    using input_type = thrust::tuple<GlobalIndexType, GlobalIndexType,
```

Comment:
The code might become easier to read if you use structs insted?

Reply:
Honestly, I'm not sure how to incorporate such a struct in thrust. Although better readability would be nice, I don't think the effort of figuring out how to use it is worth it.

Reply:
something like
```cpp
struct input {
    GlobalIndexType row;
    GlobalIndexType col;
    ValueType val;
    size_type row_range;
    size_type col_range;

    __forceinline__ __device__ __host__ input(thrust::tuple<GlobalIndexType, GlobalIndexType, ...>);
};
```

Reply:
it should probably be at file scope though, thrust doesn't like function-scoped structs.

Reply:
I think this is the only instance where this makes sense, right? The other cases use smaller tuples, so it should be easier to decipher.